File: Software/Chapter 15/cs/Listing15-4/Listing15-4/Listing15-4.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Xml;

namespace APress.Samples
{
    class XmlReaderQuery
    {
        static void Main(string[] args)
        {
            String name = "SMITH";
            String sqlstmt = "SELECT ContactID, FirstName, " +
                " COALESCE(MiddleName, '') AS MiddleName, LastName " +
                " FROM Person.Contact " +
                " WHERE LastName = @name FOR XML AUTO";
            SqlConnection sqlcon = null;
            SqlCommand sqlcmd = null;
            XmlReader sqlxr = null;
            try
            {
                sqlcon = new SqlConnection("SERVER=(local); " +
                    "INITIAL CATALOG=AdventureWorks;INTEGRATED SECURITY=SSPI;");
                sqlcon.Open();
                sqlcmd = new SqlCommand(sqlstmt, sqlcon);
                sqlcmd.Parameters.Add("@name", SqlDbType.NVarChar, 50).Value = name;
                sqlxr = sqlcmd.ExecuteXmlReader();
                while (sqlxr.Read())
                {
                    Console.WriteLine("{0}\t{1}\t{2}\t{3}",
                        sqlxr["ContactID"].ToString(),
                        sqlxr["LastName"].ToString(), 
                        sqlxr["FirstName"].ToString(), 
                        sqlxr["MiddleName"].ToString());
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                if (sqlxr != null)
                {
                    sqlxr.Close();
                }
                if (sqlcmd != null) 
                {
                    sqlcmd.Dispose();
                }
                if (sqlcon != null) 
                {
                    sqlcon.Dispose();
                }
            }
            Console.WriteLine("Press any key...");
            Console.ReadKey();
        }
    }
}

File: Software/Chapter 15/vb/Listing15-4/Listing15-4/Listing15-4.vb
Imports System.Data.SqlClient
Imports System.Xml

Namespace APress.Samples
    Module XmlReaderQuery
        Sub Main()
            Dim name As String = "SMITH"
            Dim sqlstmt As String = "SELECT ContactID, FirstName, " & _
                " COALESCE(MiddleName, '') AS MiddleName, LastName " & _
                " FROM Person.Contact " & _
                " WHERE LastName = @name FOR XML AUTO"
            Dim sqlcon As SqlConnection = Nothing
            Dim sqlcmd As SqlCommand = Nothing
            Dim sqlxr As XmlReader = Nothing
            Try
                sqlcon = New SqlConnection("SERVER=(local); " & _
                    "INITIAL CATALOG=AdventureWorks;INTEGRATED SECURITY=SSPI;")
                sqlcon.Open()
                sqlcmd = New SqlCommand(sqlstmt, sqlcon)
                sqlcmd.Parameters.Add("@name", SqlDbType.NVarChar, 50).Value = name
                sqlxr = sqlcmd.ExecuteXmlReader()
                Do While (sqlxr.Read())
                    Console.WriteLine("{0}" & ControlChars.Tab & _
                        "{1}," & ControlChars.Tab & _
                        "{2}" & ControlChars.Tab & _
                        "{3}" & ControlChars.Tab, _
                        sqlxr.Item("ContactID").ToString(), _
                        sqlxr.Item("LastName").ToString(), _
                        sqlxr.Item("FirstName").ToString(), _
                        sqlxr.Item("MiddleName").ToString())
                Loop
            Catch ex As Exception
                Console.WriteLine(ex.Message)
            Finally
                If Not (sqlxr Is Nothing) Then
                    sqlxr.Close()
                End If
                If Not (sqlcmd Is Nothing) Then
                    sqlcmd.Dispose()
                End If
                If Not (sqlcon Is Nothing) Then
                    sqlcon.Dispose()
                End If
            End Try
            Console.WriteLine("Press any key...")
            Console.ReadKey()
        End Sub
    End Module
End Namespace
